org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/DirtyStateResourceDescription.java

```diff
@@ -73,9 +75,13 @@ protected IResourceDescription createResourceDescription(Resource resource) {
 			IResourceDescription delegateDescription = delegate.getResourceDescription(resource);
 			List<IEObjectDescription> result = newArrayList();
 			String hash = getTextHash(resource);
-			ImmutableMap<String, String> userData = ImmutableMap.of(TEXT_HASH, hash);
 			for (IEObjectDescription eObjectDescription : delegateDescription.getExportedObjects()) {
-				result.add(new EObjectDescription(eObjectDescription.getQualifiedName(), eObjectDescription.getEObjectOrProxy(), userData));
+				Map<String,String> userData = new HashMap<String,String>(eObjectDescription.getUserDataKeys().length + 1);
+				for (String userDataKey : eObjectDescription.getUserDataKeys()) {
+					userData.put(userDataKey, eObjectDescription.getUserData(userDataKey));
```

Comment:
I'm not sure from the top of my head, but it could be, that this operation is blocking the entire resource set for read / write access. In that case, I'd rather like to avoid the overhead - iff type resolution doesn't happen anyway during within this lock.